keccak256/src/permutation/components.rs

```diff
@@ -0,0 +1,710 @@
+use eth_types::Field;
+use halo2_proofs::{
+    circuit::{AssignedCell, Layouter},
+    plonk::{Advice, Column, Error},
+};
+use itertools::Itertools;
+use std::convert::TryInto;
+use std::vec;
+
+use super::tables::{FromBinaryTableConfig, NUM_OF_BINARY_CHUNKS_PER_SLICE, NUM_OF_BINARY_SLICES};
+
+use crate::{
+    arith_helpers::{convert_b2_to_b13, convert_b2_to_b9, A1, A2, A3, A4, B13, B2, B9},
+    common::{NEXT_INPUTS_LANES, PERMUTATION, ROTATION_CONSTANTS, ROUND_CONSTANTS},
+    gate_helpers::{biguint_to_f, f_to_biguint},
+    permutation::{
+        generic::GenericConfig,
+        rho_helpers::{slice_lane, RhoLane},
+        tables::{
+            Base13toBase9TableConfig, FromBase9TableConfig, StackableTable,
+            NUM_OF_B9_CHUNKS_PER_SLICE, NUM_OF_B9_SLICES,
+        },
+    },
+};
+
+use num_bigint::BigUint;
+
+pub fn assign_theta<F: Field>(
+    generic: &GenericConfig<F>,
+    layouter: &mut impl Layouter<F>,
+    state: &[AssignedCell<F, F>; 25],
+) -> Result<[AssignedCell<F, F>; 25], Error> {
+    let theta_col_sums = (0..5)
+        .map(|x| {
+            generic.running_sum(
+                layouter,
+                (0..5).map(|y| state[5 * x + y].clone()).collect(),
+                None,
+            )
+        })
+        .collect::<Result<Vec<_>, Error>>()?;
+
+    let out_state = (0..5)
+        .cartesian_product(0..5)
+        .map(|(x, y)| {
+            let cells = vec![
+                state[5 * x + y].clone(),
+                theta_col_sums[(x + 4) % 5].clone(),
+                theta_col_sums[(x + 1) % 5].clone(),
+            ];
+            let vs = vec![F::one(), F::one(), F::from(B13 as u64)];
+            generic.linear_combine_consts(layouter, cells, vs, None)
+        })
+        .collect::<Result<Vec<_>, Error>>()?;
+
+    Ok(out_state.try_into().unwrap())
+}
+
+pub fn assign_rho<F: Field>(
+    layouter: &mut impl Layouter<F>,
+    base13to9_config: &Base13toBase9TableConfig<F>,
```

Comment:
If we pass the config with the columns already assigned instead of creating new ones, we are directly increasing the height of the entire thing. 

Not saying to address it now, but these are the things that are causing us to have really tall layouts.